2018/day-5/AlchemicalReduction.hs

```diff
@@ -0,0 +1,36 @@
+-- https://adventofcode.com/2018/day/5
+
+module AlchemicalReduction where
+
+import Data.Char (toLower)
+
+readLines :: String -> IO [String]
+readLines = fmap lines . readFile
+
+reducedPolymerLenght :: String -> Int
+reducedPolymerLenght = length . foldr reaction []
+
+reaction :: Char -> String -> String
+reaction x acc
```

Comment:
Am I getting it right that it returns the string reversed? - Just trying to understand some Haskell :)

Reply:
Nope, it doesn't return a reversed string. I'm not sure which part you're tying to understand, so I'll try to explain the whole idea of this solution.

So the `foldr` is very similar to Ruby's `Enum#reduce`, the first one takes a function and a list, and the latter a block and an object as its arguments. But the major difference is that `foldr` applies the passed function to each element of the list **from right to left**. So in each step it applies the function to the next element of the list and the accumulator - the result of the previous step (starting from the initial value passed to it).

Let's start with a simple example - division of each element of a list (starting from the last one) with initial value of 2.
```haskell
-- foldr functionToApply initialValue list
foldr (\) 2 [8, 12, 24, 4]

-- Step 0: We start with the whole list, a function we apply in each step and the initial value.
functionToApply = (/) -- it's a regular division
list = [8, 12, 24, 4]
previousValue = initialValue = 2

-- Step 1: foldr takes the last element of the list and divides it by the previous value.
list = [8, 12, 24]
currentValue = 4
previousValue = 2
4 / 2 = 2 -- the result is the new value of accumulator

-- Step 2
list = [8, 12]
currentValue = 24
previousValue = 2
24 / 2 = 12

-- Step 3
list = [8]
currentValue = 12
previousValue = 12
12 / 12 = 1

-- Step 4
list = [] -- the list is empty, so it's the last step
currentValue = 8
previousValue = 1
8 / 1 = 8 -- this is the final result

foldr (/) 2 [8, 12, 24, 4] == 8 -- True
```

Now we can consider the case from this puzzle. `reaction` is a function which we apply in each step of `foldr` - it's responsible for removing same letters but with different case. Let's take some polymer for the example.
```haskell
polymer = "bAcCaaD"
foldr reaction [] polymer -- an empty list is the initial value, which is equivalent to an empty string

-- Step 1
polymer = "bAcCaa"
previousValue = []
currentValue = 'D'

-- Step 2
polymer = "bAcCa"
previousValue = "D"
currentValue = 'a'

-- Step 3
polymer = "bAcC"
previousValue = "aD"
currentValue = 'a'

-- Step 4
polymer = "bAc"
previousValue = "aaD"
currentValue = 'C'

-- Step 5
polymer = "bA"
previousValue = "CaaD"
currentValue = 'c'

-- Step 6
polymer = "b"
previousValue = "aaD"
currentValue = 'A'

-- Step 7
polymer = [] -- empty list, so it's the last step
previousValue = "aD"
currentValue = 'b'

foldr reaction [] "bAcCaaD" == "baD" -- True
```

I hope the things I wrote above answers your question 😄 
Let me know if you need any further explanation.

Reply:
It does, thank you so much! I was misguided that foldr goes from left to right, not the other way around. It makes perfect sense now.